public/app/plugins/datasource/loki/live_streams.ts

```diff
@@ -42,9 +42,19 @@ export class LiveStreams {
         appendResponseToBufferedData(response, data);
         return [data];
       }),
-      catchError(err => {
-        return throwError(`error: ${err.reason}`);
-      }),
+      retryWhen((attempts: Observable<any>) =>
+        attempts.pipe(
+          mergeMap(error => {
+            // Code 1006 is used to indicate that a connection was closed abnormally.
+            // If connection was closed abnormally, we wish to retry, otherwise throw error.
+            if (error.code === 1006) {
+              console.log('Reconnecting websocket...');
+              return timer(1000);
+            }
+            return throwError(`error: ${error.reason}`);
+          })
+        )
+      ),
```

Comment:
Skilled work on the RxJs code. The code now will retry into infinity with 1s delay between attempts? Maybe we should limit the retries to x number of retries?

Reply:
Not sure if I read this right and did not find good explanation in the docs but shouldn't it return `timer(1000).take(1)` or something like that? I mean what we want is to wait for error and then wait 1s and retry and then wait for another error and repeat right? But having a timer that will fire every 1s forever after the first error does not seem correct but maybe there is some internal implementation that prevents that which isn't documented?

feels like this section from the [docs](https://www.learnrxjs.io/learn-rxjs/operators/transformation/mergemap) applies here:
> Be aware that because mergeMap maintains multiple active inner subscriptions at once it's possible to create a memory leak through long-lived inner subscriptions. A basic example would be if you were mapping to an observable with an inner timer, or a stream of dom events. In these cases, if you still wish to utilize mergeMap you may want to take advantage of another operator to manage the completion of the inner subscription, think take or takeUntil. You can also limit the number of active inner subscriptions at a time with the concurrent parameter, seen in example 5.

Reply:
Ah ok now I read that `timer(1000)` won't emit indefinitely if the second arg for period is not specified so disregard my previous comment.

Reply:
I still think we should consider adding a max number of attempts counter, but that is just me being super cautious.

Reply:
I have added emitting of warning message `Websocket connection is being disrupted. We keep reconnecting but consider starting new live tailing again.` after we reconnect 10 times and every reconnection after that. So live tailing will be still working, but it is going to be super obvious that something is wrong. Also we have changed the timer from 1s to 5s.